### ISSUE TITLE
A parser that terminates parse on failure

### ISSUE THREAD

**Issue by bonastos** [jparsec/jparsec#61]
I'm missing a parser modifier that I know from using the C++ Spirit V3 framework:

A parser that immediately terminates the parse when its contained parser fails. It may  be used to assert that the rest of an alternative matches after the begin has been matched.

As one does not have access to jparsec internals, it's impossible to implement this a clean way.
So propose to add something like `Parser<T> assertMatch()` to Parser. Of course the actual name is up to bikeshedding.

BTW, this is a hack I wrote. It seems to work ...
```java
package org.jparsec;

import org.jparsec.ParseContext;
import org.jparsec.Parser;
import org.jparsec.error.ParserException;

public class Extension {
  
  public static <T> Parser<T> assertMatch(Parser<T> parser) {
    return new Parser<T>() {
      @Override boolean apply(ParseContext ctxt) {
        if (!parser.apply(ctxt)) {
          @SuppressWarnings("deprecation")
          ParserException exception =  new ParserException(
              ctxt.renderError(), ctxt.module, ctxt.locator.locate(ctxt.errorIndex()));
          exception.setParseTree(ctxt.buildErrorParseTree());
          throw exception;
        }  
        return true;
      }
      @Override public String toString() {
        return parser.toString();
      }
    };  
  }
}
```

**Comment by fluentfuture** [comment]
Could this work?

```java
Parser<T> mustSucceed(Parser<T> parser) {
  return parser.or(Parsers.runnable(() -> {throw new MyException();}));
}
```

**Comment by bonastos** [comment]
It's  not exactly what I want. As `ScannerState.applyWithExceptionWrapped()` treats `MyException` differently from  `ParserException` , the resulting exception misses the error details. (buildParseTree instead of buildErrorParseTree) 

I'd like an exception that contains exactly the same information as the regular parse failure exception. For this the mustSucceed Parser needs internal framework details. And that's why think it should be part of the jparsec  framework.

**Comment by fluentfuture** [comment]
Can you provide some more context of your use case, especially about how the assertMatch() method helps you?

**Comment by bonastos** [comment]
Sure. I'm trying to parse a line based data file from a third party supplier. Essentially each line starts with a keyword and is followed by the data itself. Unfortunately the supplier is likely to add new keywords occasionally and we must still use other data in the file, So I came up with the following:
```
ruleX := 'KEYWORDX' defintionX EOL
unknowm := IDENTIFIER IGNORE EOL
linerule := rule1 | rule2 | ... | ruleN  | unknown
file := linerule*
```
The unknown rule will match any line with a parse error. I know that I could only match the identifier in the `unknown` rule if it's not identical to one of the known keywords. But this does triggers the parsing error in a completely different part of the grammar. It also requires quite some comments to explain to a maintainer what's going on. 

I definitely know that it's a parse error if a definition doesn't match after the keyword  has been recognized . I'd rather be able to cleanly express my intent by wrapping  a definition parser in an assertMatch  parser. 
```
ruleX := 'KEYWORDX' assertMatch(defintionX) EOL
```

**Comment by fluentfuture** [comment]
@bonastos 

Just trying to understand the use case.

You are saying that if you have a line with "GOOD_KEYWORD bad_Def", you want that line not to fail `linerule`, but instead match the unknown rule?

If you have all the line rules as separate Parser objects, and then just `or()` them, wouldn't that work?

```java
Parser lineRule = Parsers.or(rulex, ruley, ...);
```java

**Comment by bonastos** [comment]
@fluentfuture 
Sorry if I'm not clear. I have a set of keywords , each with an associated definition. Any line with a known keyword must match. If I encounter an unknown keyword I want to record it, ignore the rest of the line and continue parsing.

Let's assume I have a line that starts with the identifier 'SiTE'. The definition is an identifier. So I'd like a  to write a parser for a known line as:
``` java
  public static final Parser<String> SITE_DEF = Scanners.IDENTIFIER;
  
  public static final Parser<String> SITE_P = Scanners
      .string( "SITE" )
      .next(SITE_DEF.assertMatch())
      .followedBy(EOL);
 
```

**Comment by fluentfuture** [comment]
What breaks when you don't have `assertMatch()`?

**Comment by bonastos** [comment]
Break is maybe to hard a word, 

If a line with a known keyword is not matched, other alternatives will be tried until the last alternative is reached , which is the `unknown`  parser. 

As I do not wan't the `unknown` parser to match a known keyword, I must implement it in way that excludes all known keywords. Which means that has to be more complicated, making it more difficult to maintain. But it's certainly possible to implement.

So let's look what improves if `assertMatch()` is available. (At least in my opinion :-)

- It let's me express my intent: Fail the parse if a certain parser does not match. I specify it at the place where the parser is used.
- Each parser in the chain of alternatives is truly independent (in my special case)
- adding a parser for a new keyword is easier. (Less places to modify)

**Comment by fluentfuture** [comment]
Oh. I think I understand the problem now.

Let me rephrase so you can correct any misunderstanding I may have.

You want to write a grammar like:
```java
Parser rule1 = string("SITE").then(string("foo"));
Parser unknown = IDENTIFIER.then(...);
Parser lineRule = rule1.or(unknown);
```

And then `lineRule` should recognize both "SITE foo" and "UNKNOWN_KEYWORD whatever".
Bet the parser should still report error for "SITE bar" instead of matching it using the "unknown" rule.

Sounds about right?

Assuming that's the gist of it, I think the key is that "unknown" rule is special. Unlike rule1, rule2 which should backtrack any partial matches, "unknown" rule doesn't want to backtrack. If any previous rules has partial match, lineRule should prefer to report these partial match errors.

Adding `assertMatch()` in all of rule1, rule2, ... ruleX IMHO is inconvenient. With "unknown" rule being special, we are being forced to special-annotate all of the other rules.

We used to have an old combinator (`plus()`) that was removed in recent releases. That combinator did exactly this. It would only attempt the "alternative" parser if the previous parser had not consumed any input.

Perhaps we can resurrect that combinator (but find a less confusing name). For instance, if we call it `otherwise()`, your parser should then look like:

```java
Parser lineRule = or(rule1, rule2, rule3, ..., ruleX).otherwise(unknown);
```

What do you think?

**Comment by bonastos** [comment]
Exactly . Thanks for your persistence in trying to understand my problem.
 
`otherwise` sounds good and usable to me.

I only hope that you won't mind too much when I'll open the next issue to lobby for a LOCATION parser.

Thanks again.

**Comment by fluentfuture** [comment]
I've added otherwise() to HEAD. Please give it a try and let me know?

**Comment by abailly** [comment]
@fluentfuture It's been a while and I can't remember exactly why I had to dabble with jparsec's internals. I vaguely recall it was related to exception handling but not sure.

> The internals of jParsec isn't well thought out to be published as public API, and we may likely change the internals in future versions.

Maybe, but so what? Opening the internals could allow users to experiment with alternative designs, suggest new ideas, twist the library in unintended but nevertheless interesting ways. And there won't be a need to change the API and publish a new version each time a new valid use case crops up. 

It seems to me there are basically 2 kinds of open-source libraries: Those that provide an immediate service, that _do something_ immediately useful like providing an HTTP server, accessing a database, computing over matrices ; and those that provide building blocks for other people to build stuff, that are more like a box of Lego bricks users can recombine at will. I wish jParsec was more of the latter kind.

**Comment by bonastos** [comment]
I've checked out HEAD. The `otherwise` parser  works with simple examples, but not with my production grammar.

I'll let you know as soon as I've  found out what's going on.

**Comment by bonastos** [comment]
You beat me. I came up with the following example just before your comment arrived :-)
It shows the behaviour you described.
```java
package de.dfs.siteparser;

import static org.jparsec.Scanners.IDENTIFIER;
import static org.jparsec.Scanners.string;

import org.jparsec.Parser;
import org.jparsec.Parsers;
import org.jparsec.Scanners;

public class OtherwiseTest
{
  public static final Parser< String > UNKNOWN = IDENTIFIER.followedBy( Scanners.ANY_CHAR.skipMany() );

  public static final Parser< String > STRING_RULE = Parsers
    .or(
      string( "ABC" ).next( IDENTIFIER ),
      string( "XYZ" ).next( IDENTIFIER ) );

  public static void main(
    String[] args )
  {
    String s = null;
    //s = STRING_RULE.parse( "ABC," ); // Throws ParserException
    s = STRING_RULE.otherwise( UNKNOWN ).parse( "ABC," ); // matches fallback parser
    System.out.println( s );
  }
}
```

**Comment by fluentfuture** [comment]
Made some tweak to `otherwise()` implementation to check the `errorIndex()` instead.

Can you give it another try?

**Comment by bonastos** [comment]
works like charm.
